File: harness/lib/observability/node/logger.ts
import pino from 'pino';
import { Correlation } from './correlation';

const PII_FIELDS = [
  'password',
  'token',
  'secret',
  'authorization',
  'credit_card',
  'creditCard',
  'ssn',
  'cpf',
  'cnpj',
  'email',
  'phone',
  'phoneNumber',
];

const baseLogger = pino({
  level: process.env.LOG_LEVEL ?? 'info',
  redact: {
    paths: PII_FIELDS,
    censor: '[REDACTED]',
  },
  formatters: {
    level: (label) => ({ level: label }),
  },
  base: {
    service: process.env.OTEL_SERVICE_NAME ?? process.env.APP_NAME ?? 'app',
    env: process.env.NODE_ENV ?? 'production',
  },
  timestamp: pino.stdTimeFunctions.isoTime,
});

/**
 * Logger pré-configurado com:
 * - Redação automática de campos PII
 * - requestId e tenantId injetados automaticamente do contexto
 * - Formatação JSON estruturada
 *
 * Uso:
 *   import { logger } from '@harness/observability';
 *   logger.info({ event: 'user_created', userId }, 'User created');
 */
export const logger = {
  info(obj: Record<string, unknown>, msg?: string): void {
    baseLogger.child(getContext()).info(obj, msg);
  },
  warn(obj: Record<string, unknown>, msg?: string): void {
    baseLogger.child(getContext()).warn(obj, msg);
  },
  error(obj: Record<string, unknown> | Error, msg?: string): void {
    if (obj instanceof Error) {
      baseLogger.child(getContext()).error({ err: obj }, msg ?? obj.message);
    } else {
      baseLogger.child(getContext()).error(obj, msg);
    }
  },
  debug(obj: Record<string, unknown>, msg?: string): void {
    baseLogger.child(getContext()).debug(obj, msg);
  },
  child(bindings: Record<string, unknown>) {
    return baseLogger.child({ ...getContext(), ...bindings });
  },
};

function getContext(): Record<string, unknown> {
  const ctx: Record<string, unknown> = {};
  const requestId = Correlation.getRequestId();
  if (requestId) ctx.requestId = requestId;
  return ctx;
}

File: harness/lib/observability/python/logger.py
"""
Logger estruturado via structlog com redação automática de PII.
"""
from __future__ import annotations

import logging
import os
from typing import Any

import structlog

PII_FIELDS = {
    "password", "token", "secret", "authorization",
    "credit_card", "ssn", "cpf", "cnpj", "api_key",
}


def _redact_processor(
    _logger: Any, _method: str, event_dict: dict[str, Any]
) -> dict[str, Any]:
    for field in PII_FIELDS:
        if field in event_dict:
            event_dict[field] = "[REDACTED]"
    return event_dict


def _add_context_processor(
    _logger: Any, _method: str, event_dict: dict[str, Any]
) -> dict[str, Any]:
    from .correlation import Correlation
    request_id = Correlation.get()
    if request_id:
        event_dict["request_id"] = request_id
    event_dict.setdefault("service", os.getenv("OTEL_SERVICE_NAME", os.getenv("APP_NAME", "app")))
    event_dict.setdefault("env", os.getenv("PYTHON_ENV", "production"))
    return event_dict


def configure_logging(level: str = "INFO") -> None:
    logging.basicConfig(level=getattr(logging, level.upper(), logging.INFO))
    structlog.configure(
        processors=[
            structlog.contextvars.merge_contextvars,
            structlog.stdlib.add_log_level,
            structlog.stdlib.add_logger_name,
            structlog.processors.TimeStamper(fmt="iso"),
            _redact_processor,
            _add_context_processor,
            structlog.processors.StackInfoRenderer(),
            structlog.processors.ExceptionRenderer(),
            structlog.processors.JSONRenderer(),
        ],
        wrapper_class=structlog.stdlib.BoundLogger,
        context_class=dict,
        logger_factory=structlog.stdlib.LoggerFactory(),
    )


configure_logging(os.getenv("LOG_LEVEL", "INFO"))
logger = structlog.get_logger()
